Subnet deploy prod (#949)
ava-labs/avalanche-docs: docs/subnets/deploying-subnets-on-prod.md

* [Amazon Web Services](../nodes/build/setting-up-an-avalanche-node-with-amazon-web-services-aws.md)
* [Azure](../nodes/build/set-up-an-avalanche-node-with-microsoft-azure.md)
* [Google Cloud Platform](../nodes/build/set-up-an-avalanche-node-with-google-cloud-platform.md)

There is a whole range of other cloud providers that may offer lower prices or better deals for your particular needs, so it makes sense to shop around.

Once you decide on a provider (or providers), if they offer instances in multiple data centers, it makes sense to spread the nodes geographically since that provides a better resilience and stability against outages.

### Number of Validators

Number of validators on a subnet is a crucial decision you need to make. For stability and decentralization, you should strive to have as many validators as possible.

For stability reasons our recommendation is to have **at least** 5 full validators on your Subnet. If you have less than 5 validators your Subnet liveness will be at risk whenever a single validator goes offline, and if you have less than 4 even one offline node will halt your Subnet.

> Considering that at times you will have to take nodes offline, for routine maintenance (at least for node upgrades which happen with some regularity) or unscheduled outages and failures you need to be able to routinely handle at least one node being offline without your Subnet performance degrading.

Reviewer: Do we want to recommend more than 5?

Author: As far as I understand, 5 is the bare minimum from a subnet stability standpoint. The more, the better, but since running nodes is a non-trivial expense (2000AVAX per validator) I would hesitate to make more than 5 an official recommendation. I will add language to ephasize that more validators is certainly better. If the platform team feels that 5 is not enough for production Subnets, we can change it. Let me know.

### Node bootstrap

Once you set up the node instances and install AvalancheGo clients on them, nodes will need to boostrap (sync with the network). This is a lengthy process, as the nodes need to catch up and replay all the network activity since the genesis up to the present moment. Full bootstrap on a node can take more than a week, but there are ways to shorten that process, depending on your circumstances.

#### State sync

> If the nodes you will be running as validators don't need to have the full transaction history, then you can use [state sync](../nodes/maintain/chain-config-flags.md#state-sync-enabled-boolean). With this flag enabled, instead of replaying the whole history to get to the current state, nodes simply download only the current state from other network peers, shortening the bootstrap process from multiple days to a couple of hours. If the nodes will be used for subnet validation exclusively, you can use the state sync without any issues.

Reviewer: Note: we should specify state sync is just on the C-Chain

Author: Good call, will do.

To have a node start syncing the Subnet, you need to add the `--whitelisted-subnets` command line option, or `whitelisted-subnets` key to the node config file (found at `.avalanchego/configs/node.json` for installer-script created nodes). A single node can sync multiple subnets, so you can add them as a comma-separated list of subnet IDs.

An example of a node config syncing two subnets:

```json
{
  "dynamic-public-ip": "opendns",
  "http-host": "",
  "whitelisted-subnets": "28nrH5T2BMvNrWecFcV3mfccjs6axM1TVyqe79MCv2Mhs8kxiY,Ai42MkKqk8yjXFCpoHXw7rdTWSHiKEMqh5h8gbxwjgkCUfkrk"
}
```

But that is not all. Besides the whitelist containing the SubnetID, node also needs to have the plugin that contains the VM instance the blockchain in the Subnet will run. You should have already been through that on testnet and Fuji, but for a refresher, you can refer to [this tutorial](create-a-fuji-subnet.md).

> So, name the VM plugin binary as the `VMID` of the Subnet chain and place it in the `plugins` directory where the node binary is (for installer-script created nodes that would be `~/avalanche-node/plugins/`). 

Reviewer: It seems a little strange to reference this here. I guess this is more of a "considerations for deploying on production" document than a walk through tutorial, but if we want this to be a walkthrough or have a section that does the full walkthrough, I think we should explicitly state in that walkthrough to make sure the needed VM binaries are in place.

Author: To make this an explicit walkthrough we would need to duplicate a lot of material from other tutorials (installation, subnet-cli and so on) and we wanted to avoid that (tools are rapidly evolving and copies get outdated, maintenance is messy). Instead we opted for linking to source material as that will be up to date.

Explicit mention of the plugin binary comes from experience, lots of people stumble on that detail, but I'm ok with removing it if you feel strongly about it.

If you want to inspect the process of Subnet syncing, you can use the RPC call to check for the [blockchain status](../apis/avalanchego/apis/p-chain.md#platformgetblockchainstatus).

For a more in-depth look into Subnet operation, check out the blockchain log. By default, the log can be found in `~/.avalanchego/logs/ChainID.log` where you replace the `ChainID` with the actual ID of the blockchain in your Subnet.

For an even more thorough (and pretty!) insight into how the node and the Subnet is behaving, you can install the Prometheus+Grafana monitoring system with the custom dashboards for the regular node operation, as well as a dedicated dashboard for Subnet data. Check out the [tutorial](../nodes/maintain/setting-up-node-monitoring.md) for information on how to set it up.

### Managing validation

On Avalanche all validations are limited in time and can range from two weeks up to one year. Furthermore, subnet validations are always a subset of the Primary Network validation period (must be shorter or the same). That means that periodically your validators will expire and you will need to submit a new validation transaction for both the Primary Network and your Subnet.

> Unless managed properly and in a timely manner, that can be disruptive for your Subnet (if all validators expire at the same time your Subnet will halt). To avoid that, keep notes on when a particular validation is set to expire and be ready to renew it as soon as possible. Also, when initially setting up the nodes, make sure to stagger the validator expiry so they don't all expire on the same date. Setting end dates at least a day apart is a good practice, as well as setting reminders for each expiry.

Reviewer: 🔥

## Conclusion

Hopefully, by reading this document you have a better picture of the requirements and considerations you need to make when deploying your subnet to production and you are now better prepared to launch your Subnet successfully.

> Keep in mind, running a Subnet in production is not a one-and-done kind of situation, it is in fact running a fleet of servers 24/7. Things can change and go sideways in unexpected ways so you always need to be on your toes, and constantly monitor the nodes and subnet health. 

Reviewer: Can we change the gist of this from "this could break in any way" to, this is a running fleet of servers that requires constant monitoring as with any critical real time service?

Author: Will do.

# Deploying Subnets on Production Infrastructure

## Introduction

> After architecting your Subnet environment on the [local machine](create-a-local-subnet.md), proving the design and testing it out on [the testnet](create-a-fuji-subnet.md), eventually you will need to deploy your Subnet to production environment. Running a subnet in production is much more involved than local and testnet deploys, as your Subnet will have to take care of real world usage, maintaining uptime, upgrades and all of that in a potentially adversarial environment. The purpose of this document is to point out a set of general considerations and propose potential solutions to them.

Reviewer 2: ```suggestion
After architecting your Subnet environment on the [local machine](create-a-local-subnet.md), proving the design and testing it out on [the testnet](create-a-fuji-subnet.md), eventually you will need to deploy your Subnet to production environment. Running a Subnet in production is much more involved than local and testnet deploys, as your Subnet will have to take care of real world usage, maintaining uptime, upgrades and all of that in a potentially adversarial environment. The purpose of this document is to point out a set of general considerations and propose potential solutions to them.
```
Please search and replace all `subnet` to `Subnet`

Author: I capitalized Subnet when talking about a particular subnet (your Subnet). Left with lowercase when talking about generic subnets. If the style guide is to capitalize subnets wholesale, I can correct.

Reviewer 2: Marketing said to capitalize subnets

The architecture of the environment your particular Subnet will use will be greatly influenced by the type of load and activity your Subnet is designed to support so your solution will most likely differ from what we propose here. Still, it might be useful to follow along, to build up the intuition for the type of questions you will need to consider.

> ## Node setup

Reviewer 2: Section headers should be Title Case: https://titlecase.com/, please update all section headers.

> Nodes are essential elements for running your Subnet in production. At a minimum, your Subnet will need validator nodes, potentially also nodes that act as RPC servers, indexers or explorers. Running a node is basically running an instance of [AvalancheGo](https://github.com/ava-labs/avalanchego) on a server.

Reviewer 2: ```suggestion
Avalanche nodes are essential elements for running your Subnet in production. At a minimum, your Subnet will need validator nodes, potentially also nodes that act as RPC servers, indexers or explorers. Running a node is basically running an instance of [AvalancheGo](../nodes/README.md) on a server.
```

### Server OS

Although AvalancheGo can run on a MacOS or a Windows computer, we strongly recommend running nodes on computers running Linux as they are designed specifically for server loads and all the tools and utilities needed for administering a server are native to Linux.

### Hardware specification

For running AvalancheGo as a validator on the Primary Network the recommended configuration is as follows:

* CPU: Equivalent of 8 AWS vCPU
* RAM: 16 GiB
* Storage: 1TB with at least 3000 IOPS
* OS: Ubuntu 18.04/20.04
* Network: Reliable IPv4 or IPv6 network connection, with an open public port

That is the configuration sufficient for running a Primary Network node. Any resource requirements for your Subnet come on top of this, so you should not go below this configuration, but may need to step up the specification if you expect your Subnet to handle a significant amount of transactions.

> Be sure to set up monitoring of resource consumption for your nodes because resource exhaustion may cause your Subnet to slow down or even halt.

Reviewer 2: ```suggestion
Be sure to set up monitoring of resource consumption for your nodes because resource exhaustion may cause your node to slow down or even halt, which may severely impact your Subnet negatively.
```

Author: Yep, good call.

### Number of Validators

Number of validators on a subnet is a crucial decision you need to make. For stability and decentralization, you should strive to have as many validators as possible.

For stability reasons our recommendation is to have **at least** 5 full validators on your Subnet. If you have less than 5 validators your Subnet liveness will be at risk whenever a single validator goes offline, and if you have less than 4 even one offline node will halt your Subnet.

You should be aware that 5 is the minimum we recommend. But, from a decentralization standpoint having more validators is always better as it increases the stability of your Subnet and makes it more resilient to both technical failures and adversarial action. In a nutshell: run as many Subnet validators as you can.

Considering that at times you will have to take nodes offline, for routine maintenance (at least for node upgrades which happen with some regularity) or unscheduled outages and failures you need to be able to routinely handle at least one node being offline without your Subnet performance degrading.

### Node bootstrap

> Once you set up the node instances and install AvalancheGo clients on them, nodes will need to boostrap (sync with the network). This is a lengthy process, as the nodes need to catch up and replay all the network activity since the genesis up to the present moment. Full bootstrap on a node can take more than a week, but there are ways to shorten that process, depending on your circumstances.

Author: Will change to 'node'.

#### State sync

If the nodes you will be running as validators don't need to have the full transaction history, then you can use [state sync](../nodes/maintain/chain-config-flags.md#state-sync-enabled-boolean). With this flag enabled, instead of replaying the whole history to get to the current state, nodes simply download only the current state from other network peers, shortening the bootstrap process from multiple days to a couple of hours. If the nodes will be used for subnet validation exclusively, you can use the state sync without any issues. Currently, state sync is only available for the C-Chain, but since the bulk of the transactions on the platform happen there it still has a significant impact on the speed of bootstrapping.

#### Database copy

> Good way to cut down on bootstrap times on multiple nodes is database copy. Database is identical across nodes, and as such can safely be copied from one node to another. Just make sure to that the node is not running during the copy process, as that can result in a corrupted database. Database copy procedure is explained in detail [here](../nodes/maintain/node-backup-and-restore.md#database).

Reviewer 2: ```suggestion
Good way to cut down on bootstrap times on multiple nodes is database copy. Database is identical across nodes, and as such can safely be copied from one node to another. Just make sure to that the node is not running during the copy process, as that can result in a corrupted database. Database copy procedure is explained in detail [here](../nodes/maintain/node-backup-and-restore.md#database).

Please make sure you don't reuse any node's node-id by accident, especially don't restore another node's ID, see [here](../nodes/maintain/node-backup-and-restore.md#nodeid) for details. Each node must has its own unique node id, otherwise, the nodes sharing the same id will not behave correctly, which will impact your validator's uptime, thus staking rewards, and the stability of your Subnet.
```

Author: Yep, grea suggestion.

## Subnet deploy

> Once you have the nodes set up you are ready to deploy the actual subnet. Right now, the recommended tool to do that is [Subnet-CLI](https://github.com/ava-labs/subnet-cli).

Reviewer 2: ```suggestion
Once you have the nodes set up you are ready to deploy the actual Subnet. Right now, the recommended tool to do that is [Subnet-CLI](https://github.com/ava-labs/subnet-cli).
```

Instructions for using the Subnet-CLI can be found [here](../subnets/subnet-cli.md). We will highlight the main steps below.

### Ledger HW wallet

> When creating the Subnet, you will be required to have a private key that will control the administrative functions of the subnet (adding validators, managing the configuration). Needless to say, whoever has this private key has complete control over the Subnet and the way it runs. Therefore, protecting that key is of the utmost operational importance. Which is why we strongly recommend using a hardware wallet such as a [Ledger HW Wallet](https://www.ledger.com/) to store and access that private key. 

Reviewer 2: ```suggestion
When creating the Subnet, you will be required to have a private key that will control the administrative functions of the Subnet (adding validators, managing the configuration). Needless to say, whoever has this private key has complete control over the Subnet and the way it runs. Therefore, protecting that key is of the utmost operational importance. Which is why we strongly recommend using a hardware wallet such as a [Ledger HW Wallet](https://www.ledger.com/) to store and access that private key. 
```

Of course, Subnet-CLI supports the usage of a Ledger HW wallet. Take advantage of that because losing control of the managing would be catastrophic. General instruction on how to use a Ledger device with Avalanche can be found [here](https://support.avax.network/en/articles/6150237-how-to-use-a-ledger-nano-s-or-nano-x-with-avalanche).

### Genesis file

> The structure that defines the most important parameters in a subnet is found in the genesis file, which is a `json` formatted, human-readable file. Describing the contents and the options available in the genesis file is beyond the scope of this document, and if you're ready to deploy your Subnet to production you probably have it mapped out already. 

Reviewer 2: ```suggestion
The structure that defines the most important parameters in a Subnet is found in the genesis file, which is a `json` formatted, human-readable file. Describing the contents and the options available in the genesis file is beyond the scope of this document, and if you're ready to deploy your Subnet to production you probably have it mapped out already. 
```

If you want to review, we have a description of the genesis file in our document on [customizing EVM Subnets](customize-a-subnet.md).

### Subnet-CLI Wizard

> Creating a subnet is a multistep process. You need to:

Reviewer 2: ```suggestion
Creating a Subnet is a multistep process. You need to:
```

An example of wizard command line could look like:

```bash
subnet-cli wizard \
--ledger \
--node-ids=NodeID-741aqvs6R4iuHDyd1qT1NrFTmsgu78dc4,NodeID-K7Y79oAmBntAcdkyY1CLxCim8QuqcZbBp,NodeID-C3EY6u4v7DDi6YEbYf1wmXdvkEFXYuXNW,NodeID-AiLGeqQfh9gZY3Y8wLMD15tuJtsJHq5Qi \
--vm-genesis-path=prod-genesis.json \
--vm-id=tGas3T58KzdjLHhBDMnH2TvrddhqTji5iZAMZ3RXs2NLpSnhH \
--chain-name=prodSubnet
```

Please refer to the Subnet-CLI docs for [detailed instructions](subnet-cli.md#subnet-cli-wizard).

If the wizard completes correctly your Subnet will now be created, deployed and available to see in [the Subnet explorer](https://subnets.avax.network/subnets). Congrats!

You can also use Subnet-CLI to manage your Subnet after the initial deploy, to add new validators, renew ones that expired and so on. Familiarize yourself with Subnet-CLI, you be using it often!

## Validator Configuration

Running nodes as Subnet validators warrants some additional considerations, above those when running a regular node or a Primary Network-only validator.

### Subnet Whitelisting

For a node to join a subnet, there are two prerequisites:
* Primary Network validation
* Subnet whitelisting

> Primary Network validation means that a node cannot join a subnet as a validator before becoming a validator on the Primary Network itself. So, after you add the node to the validator set on the Primary Network, node can join a subnet. Of course, this is valid only for subnet validators, if you need a non-validating subnet node, then the node doesn't need to be a validator at all.

Reviewer 2: ```suggestion
Primary Network validation means that a node cannot join a Subnet as a validator before becoming a validator on the Primary Network itself. So, after you add the node to the validator set on the Primary Network, node can join a Subnet. Of course, this is valid only for Subnet validators, if you need a non-validating Subnet node, then the node doesn't need to be a validator at all.
```

}
```

But that is not all. Besides the whitelist containing the SubnetID, node also needs to have the plugin that contains the VM instance the blockchain in the Subnet will run. You should have already been through that on testnet and Fuji, but for a refresher, you can refer to [this tutorial](create-a-fuji-subnet.md).

So, name the VM plugin binary as the `VMID` of the Subnet chain and place it in the `plugins` directory where the node binary is (for installer-script created nodes that would be `~/avalanche-node/plugins/`). 

### Subnet Bootstrapping

After you have added the whitelist and placed the VM binary in the correct directory, your node is ready to start syncing with the Subnet. Restart the node and monitor the log output. You should notice something similar to:

```text
Jul 30 18:26:31 node-fuji avalanchego[1728308]: [07-30|18:26:31.422] INFO chains/manager.go:262 creating chain:
Jul 30 18:26:31 node-fuji avalanchego[1728308]:     ID: 2ebCneCbwthjQ1rYT41nhd7M76Hc6YmosMAQrTFhBq8qeqh6tt
Jul 30 18:26:31 node-fuji avalanchego[1728308]:     VMID:srEXiWaHuhNyGwPUi444Tu47ZEDwxTWrbQiuD7FmgSAQ6X7Dy
```

> That means the node has detected the subnet, and is attempting to initialize it and start bootstrapping the Subnet. It might take some time (if there are already transactions on the subnet), and eventually it will finish the bootstrap with a message like:

Reviewer 2: ```suggestion
That means the node has detected the subnet, and is attempting to initialize it and start bootstrapping the Subnet. It might take some time (if there are already transactions on the Subnet), and eventually it will finish the bootstrap with a message like:
```

### Managing validation

On Avalanche all validations are limited in time and can range from two weeks up to one year. Furthermore, subnet validations are always a subset of the Primary Network validation period (must be shorter or the same). That means that periodically your validators will expire and you will need to submit a new validation transaction for both the Primary Network and your Subnet.

Unless managed properly and in a timely manner, that can be disruptive for your Subnet (if all validators expire at the same time your Subnet will halt). To avoid that, keep notes on when a particular validation is set to expire and be ready to renew it as soon as possible. Also, when initially setting up the nodes, make sure to stagger the validator expiry so they don't all expire on the same date. Setting end dates at least a day apart is a good practice, as well as setting reminders for each expiry.

## Conclusion

> Hopefully, by reading this document you have a better picture of the requirements and considerations you need to make when deploying your subnet to production and you are now better prepared to launch your Subnet successfully.

Reviewer 2: ```suggestion
Hopefully, by reading this document you have a better picture of the requirements and considerations you need to make when deploying your Subnet to production and you are now better prepared to launch your Subnet successfully.
```

> Keep in mind, running a Subnet in production is not a one-and-done kind of situation, it is in fact running a fleet of servers 24/7. And as with any real time service, you should have a robust logging, monitoring and alerting systems to constantly check the nodes and subnet health and alert you if anything out of the ordinary happens. 

Reviewer 2: ```suggestion
Keep in mind, running a Subnet in production is not a one-and-done kind of situation, it is in fact running a fleet of servers 24/7. And as with any real time service, you should have a robust logging, monitoring and alerting systems to constantly check the nodes and Subnet health and alert you if anything out of the ordinary happens. 
```

#### Local Servers

If you plan on running nodes on your own hardware, make sure they satisfy the minimum HW specification as outlined earlier. Pay close attention to proper networking setup, making sure the p2p port (9651) is accessible and public IP properly configured on the node. Make sure the node is connected to the network physically (not over WiFi), and that the router is powerful enough to handle a couple of thousands of persistent TCP connections and that network bandwidth can accommodate at least 5Mbps of steady upstream and downstream network traffic.

When installing the AvalancheGo node on the machines, unless you have a dedicated DevOps staff that will take care of node setup and configuration, we recommend using the [installer script](../nodes/build/set-up-node-with-installer.md) to set up the nodes. It will abstract most of the setup process for you, set up the node as a system service and will enable easy node upgrades.

#### Cloud Providers

There are a number of different cloud providers. We have documents that show how to set up a node on the most popular ones:

> * [Amazon Web Services](../nodes/build/setting-up-an-avalanche-node-with-amazon-web-services-aws.md)

Reviewer 3: [MD032/blanks-around-lists: Lists should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md032---lists-should-be-surrounded-by-blank-lines)

Of course, Subnet-CLI supports the usage of a Ledger HW wallet. Take advantage of that because losing control of the managing would be catastrophic. General instruction on how to use a Ledger device with Avalanche can be found [here](https://support.avax.network/en/articles/6150237-how-to-use-a-ledger-nano-s-or-nano-x-with-avalanche).

### Genesis File

The structure that defines the most important parameters in a Subnet is found in the genesis file, which is a `json` formatted, human-readable file. Describing the contents and the options available in the genesis file is beyond the scope of this document, and if you're ready to deploy your Subnet to production you probably have it mapped out already. 

If you want to review, we have a description of the genesis file in our document on [customizing EVM Subnets](customize-a-subnet.md).

### Subnet-CLI Wizard

Creating a Subnet is a multistep process. You need to:

> * create a `WMID`

Reviewer 3: [MD032/blanks-around-lists: Lists should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md032---lists-should-be-surrounded-by-blank-lines)

An example of wizard command line could look like:

```bash
subnet-cli wizard \
--ledger \
--node-ids=NodeID-741aqvs6R4iuHDyd1qT1NrFTmsgu78dc4,NodeID-K7Y79oAmBntAcdkyY1CLxCim8QuqcZbBp,NodeID-C3EY6u4v7DDi6YEbYf1wmXdvkEFXYuXNW,NodeID-AiLGeqQfh9gZY3Y8wLMD15tuJtsJHq5Qi \
--vm-genesis-path=prod-genesis.json \
--vm-id=tGas3T58KzdjLHhBDMnH2TvrddhqTji5iZAMZ3RXs2NLpSnhH \
--chain-name=prodSubnet
```

Please refer to the Subnet-CLI docs for [detailed instructions](subnet-cli.md#subnet-cli-wizard).

If the wizard completes correctly your Subnet will now be created, deployed and available to see in [the Subnet explorer](https://subnets.avax.network/subnets). Congrats!

You can also use Subnet-CLI to manage your Subnet after the initial deploy, to add new validators, renew ones that expired and so on. Familiarize yourself with Subnet-CLI, you be using it often!

## Validator Configuration

Running nodes as Subnet validators warrants some additional considerations, above those when running a regular node or a Primary Network-only validator.

### Subnet Whitelisting

For a node to join a subnet, there are two prerequisites:

> * Primary Network validation

Reviewer 3: [MD032/blanks-around-lists: Lists should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md032---lists-should-be-surrounded-by-blank-lines)

Creating a Subnet is a multistep process. You need to:

> - create a `VMID`

Reviewer 2: Good catch
